### PR TITLE
State that external objects must not depend on themselves.

### LINKED ISSUE #2845: Add non-cyclic rule for external object construction
This is a follow-up to #1907.  As I understand the intention of that discussion, we want to avoid entangled external object construction.  At the same time, I expect that we want more freedom than having to construct all external objects even before solving the initialization problem.

What I think we want is something along these lines (perhaps right after the sentence with _called exactly once before the first use of the object_):
> The arguments to the constructor must not – directly nor indirectly – depend on the external object being constructed.

Here's an example of a cyclic dependency that would be illegal:
```
model CyclicExternalObjectConstruction

  class EO
    extends ExternalObject;
    …
  end EO;

  function get
    input EO obj;
    output Real y;
    …
  end get;

  EO obj = EO(x);
  parameter Real x = get(obj);

end CyclicExternalObjectConstruction;
```

### PR DESCRIPTION
**Pull request by HansOlsson**
Closes #2845

### CODE REVIEW

File: chapters/functions.tex [hunk 1]
```diff
@@ -2470,8 +2470,9 @@ \subsection{External Objects}\label{external-objects}
   \lstinline!ExternalObject!. It is not legal to call explicitly the \lstinline!constructor! and
   \lstinline!destructor! functions. The constructor shall initialize the object, and
   must not require any other calls to be made for the initialization to
-  be complete (e.g., from an initial algorithm or initial equation). The
-  destructor shall delete the object, and must not require any other
+  be complete (e.g., from an initial algorithm or initial equation).
+  The arguments to the constructor must not – directly nor indirectly – depend on the external object being constructed.
```
Comment (henrikt-ma) [inline chapters/functions.tex:2474]: Use non-unicode en-dash:
```suggestion
  The arguments to the constructor must not -- directly nor indirectly -- depend on the external object being constructed.
```

**Review by henrikt-ma — CHANGES_REQUESTED** [review]
Apart from the copy/paste error with the dashes, this would work.

However, I still think this rule is more important than the details about the constructor and destructor, and that it would be more natural to add it as the second sentence of the list item.

Then, slightly off-topic, should we also fix the ordering of the words _not legal to call explicitly the_ → _not legal to explicitly call the_?

**Comment by HansOlsson** [comment]
> Apart from the copy/paste error with the dashes, this would work.
> 
> However, I still think this rule is more important than the details about the constructor and destructor, and that it would be more natural to add it as the second sentence of the list item.

To me it seemed to fit better in this place, since I found it natural that we first say that it constructs it completely, and then that there cannot be cyclic dependencies.

> Then, slightly off-topic, should we also fix the ordering of the words _not legal to call explicitly the_ → _not legal to explicitly call the_?

Probably. Would be good with a more complete review of that.

**Comment by henrikt-ma** [comment]
> To me it seemed to fit better in this place, since I found it natural that we first say that it constructs it completely, and then that there cannot be cyclic dependencies.

OK, no big deal.  It can probably be thought of as a chicken-and-egg problem; there is probably a perspective that I'm missing.  To me, speaking of what happens during construction (constructing it completely) before speaking of what is required to even call the constructor seemed backward.

Actually, the whole list item is lengthy and rather poorly organized.  Here is a more complete proposal for the description of `constructor` and `destructor`:
```
\item
  An external object class shall be directly extended from \lstinline!ExternalObject!, shall have exactly two function definitions, called \lstinline!constructor! and \lstinline!destructor!, and shall not contain other elements.
  The functions \lstinline!constructor! and \lstinline!destructor! shall not be replaceable.
  It is not legal to explicitly call the \lstinline!constructor! and \lstinline!destructor! functions.
\item
  The \lstinline!constructor! function is called exactly once before the first use of the object.
  The \lstinline!constructor! shall have exactly one output argument in which the constructed instance derived from \lstinline!ExternalObject! is returned.
  The arguments to the constructor must not -- directly nor indirectly -- depend on the external object being constructed.
  The constructor shall initialize the object, and must not require any other calls to be made for the initialization to be complete (e.g., from an initial algorithm or initial equation).

  The constructor shall not assume that pointers sent to the external object will remain valid for the life-time of the external object.
  An exception is that if the pointer to another external object is given as argument to the constructor, that pointer will remain valid as long as the other external object lives.
\item
  For each completely constructed object, the \lstinline!destructor! is called exactly once, after the last use of the object, even if an error occurs.
  The \lstinline!destructor! shall have no output arguments and the only input argument of the destructor shall be of the type derived from \lstinline!ExternalObject!.
  The destructor shall delete the object, and must not require any other calls to be made for the deletion to be complete (e.g., from a \lstinline!when terminal()! clause).
  \begin{nonnormative}
  External objects may be a protected component (or part of one) in a function.
  The constructor is in that case called at the start of the function call, and the destructor when the function returns, or when recovering from errors in the function.
  \end{nonnormative}
  \begin{nonnormative}
  External objects may be an input (or part of an input) to a function, in that case the destructor is not called (since the external object is active before and after the function call).
  Normally this is an external function, but it could be a non-external function as well (e.g.\ calling external functions one or more times).
  The function input shall not have a default value using the constructor.
  \end{nonnormative}
```

**Comment by henrikt-ma** [comment]
I also almost forgot my new responsibility as reviewer: I think this deserves being mentioned in change log, so we should probably work out how to handle the change log before closing this PR.

I should mention that I'll be away next week, so don't expect any activity on my side until March 1.

**Comment by HansOlsson** [comment]
> I also almost forgot my new responsibility as reviewer: I think this deserves being mentioned in change log, so we should probably work out how to handle the change log before closing this PR.

Although I agree that it is important, I think we can do that later.
There are already some PRs closed on master after 3.5 that we will need to handle.

**Review by henrikt-ma — APPROVED** [review]
OK.  I'll open a separate PR with the restructuring changes, so that this can be closed.

**Pull request merged**